src/test/java/ArrayIndexedBagTest.java

```diff
@@ -0,0 +1,449 @@
+        bag.add(101);
+        bag.remove(Integer.valueOf(101));
+        assertNotEquals(101, bag.get(1));
+    }
+
+    @Test
+    void removeByValueForIndexedBagUpdatesSize() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        bag.add(99);
+        bag.add(101);
+        bag.add(103);
+        bag.remove(Integer.valueOf(101));
+        assertEquals(2, bag.size());
+    }
+
+    @Test
+    void removeByValueSoleElementForIndexedBagIsEmpty() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        bag.add(99);
+        bag.remove(Integer.valueOf(99));
+        assertTrue(bag.isEmpty());
+    }
+
+    @Test
+    void removeByValueOfNonExistingElementInIndexedBagThrowsException() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        bag.add(99);
+        bag.add(101);
+        bag.add(103);
+        assertThrows(NoSuchElementException.class, () -> bag.remove(Integer.valueOf(110)));
+    }
+
+    @Test
+    void removeByIndexFromIndexedBagReturnsValue() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        bag.add(99);
+        bag.add(101);
+        bag.add(103);
+        assertEquals(101, bag.remove(1));
+    }
+
+    @Test
+    void removeByIndexRemovesElementFromIndexedBag() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        bag.add(99);
+        bag.add(101);
+        bag.add(103);
+        bag.remove(1);
+        assertFalse(bag.contains(101));
+    }
+
+    @Test
+    void removeByIndexOfIndexedBagUpdatesSize() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        bag.add(99);
+        bag.add(101);
+        bag.add(103);
+        bag.remove(1);
+        assertEquals(2, bag.size());
+    }
+
+    @Test
+    void removeByIndexSoleElementForIndexedBagIsEmpty() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        bag.add(99);
+        bag.remove(0);
+        assertTrue(bag.isEmpty());
+    }
+
+    @Test
+    void removeByIndexNotInRangeOfIndexedBagThrowsException() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        bag.add(99);
+        bag.add(101);
+        bag.add(103);
+        assertThrows(IndexOutOfBoundsException.class, () -> bag.remove(22));
+    }
+
+    @Test
+    void removeByNegativeIndexOfIndexedBagPropagatesException() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        assertThrows(IndexOutOfBoundsException.class, () -> bag.remove(-1));
+    }
+
+    @Test
+    void iteratorReturnsIterator() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        Iterator<Integer> it = bag.iterator();
+        assertNotNull(it);
+        assertTrue(it instanceof Iterator);
+    }
+
+    @Test
+    void iteratorHasNextOnNonEmptyIndexedBagReturnsTrue() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        bag.add(99);
+        Iterator<Integer> it = bag.iterator();
+        assertTrue(it.hasNext());
+        assertTrue(it.hasNext());
+    }
+
```

Comment:
Worth making this two seperate tests?
(a) has next is true
(b) has next stays true if no `next`